File: backend/app/models/portal_access_keys.py
# Copyright (c) 2026 Laurent Barbe
# Licensed under the Apache License, Version 2.0

"""Portal access-key API contracts."""

from datetime import datetime
from typing import Literal, Optional

from pydantic import Field, field_validator

from app.models.base import ApiModel


class PortalAccessKey(ApiModel):
    access_key_id: str
    status: Optional[str] = None
    created_at: Optional[str] = None
    is_active: bool = False
    is_portal: bool = False
    deletable: bool = True
    secret_access_key: Optional[str] = None
    expires_at: Optional[datetime] = None
    session_token: Optional[str] = None
    target_type: Literal["self", "external"] = "self"
    external_email: Optional[str] = None
    storage_space_name: Optional[str] = None
    bucket_name: Optional[str] = None
    permission: Optional[Literal["read_only", "read_write"]] = None


class PortalAccessKeyCreate(ApiModel):
    target_type: Literal["self", "external"] = "self"
    storage_space_id: Optional[str] = Field(default=None, min_length=1, max_length=1024)
    external_email: Optional[str] = Field(default=None, max_length=254)
    permission: Optional[Literal["read_only", "read_write"]] = None

    @field_validator("storage_space_id")
    @classmethod
    def _validate_storage_space_id(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Storage Space is required")
        return cleaned

    @field_validator("external_email")
    @classmethod
    def _validate_external_email(cls, value: Optional[str]) -> Optional[str]:
        if value is None:
            return None
        cleaned = " ".join(value.split())
        if not cleaned:
            raise ValueError("External user label is required")
        return cleaned


class PortalAccessKeyStatusChange(ApiModel):
    active: bool


class PortalIAMUser(ApiModel):
    iam_user_id: Optional[str] = None
    iam_username: Optional[str] = None
    arn: Optional[str] = None
    created_at: Optional[datetime] = None


class PortalAccessKeysState(ApiModel):
    iam_user: PortalIAMUser
    s3_endpoint: Optional[str] = None
    force_path_style: bool = False
    access_keys: list[PortalAccessKey]
    can_manage_access_keys: bool = False
    max_access_keys: int = Field(default=2, ge=1)

File: frontend/src/api/portalAccessKeys.ts
/*
 * Copyright (c) 2026 Laurent Barbe
 * Licensed under the Apache License, Version 2.0
 */
import { S3AccountSelector, withS3AccountParam } from "./accountParams";
import client from "./client";

export type PortalAccessKey = {
  access_key_id: string;
  status?: string | null;
  created_at?: string | null;
  is_active: boolean;
  is_portal?: boolean;
  deletable?: boolean;
  secret_access_key?: string | null;
  session_token?: string | null;
  expires_at?: string | null;
  target_type?: "self" | "external";
  external_email?: string | null;
  storage_space_name?: string | null;
  bucket_name?: string | null;
  permission?: "read_only" | "read_write" | null;
};

export type PortalAccessKeyCreate = {
  target_type?: "self" | "external";
  storage_space_id?: string | null;
  external_email?: string | null;
  permission?: "read_only" | "read_write" | null;
};

export type PortalIAMUser = {
  iam_user_id?: string | null;
  iam_username?: string | null;
  arn?: string | null;
  created_at?: string | null;
};

export type PortalAccessKeysState = {
  iam_user: PortalIAMUser;
  s3_endpoint?: string | null;
  force_path_style?: boolean;
  access_keys: PortalAccessKey[];
  can_manage_access_keys: boolean;
  max_access_keys: number;
};

export async function fetchPortalAccessKeysState(
  accountId: S3AccountSelector,
): Promise<PortalAccessKeysState> {
  const { data } = await client.get<PortalAccessKeysState>("/portal/access-keys", {
    params: withS3AccountParam(undefined, accountId),
  });
  return data;
}

export async function createPortalAccessKey(
  accountId: S3AccountSelector,
  payload?: PortalAccessKeyCreate,
): Promise<PortalAccessKey> {
  const { data } = await client.post<PortalAccessKey>(
    "/portal/access-keys",
    payload,
    { params: withS3AccountParam(undefined, accountId) },
  );
  return data;
}

export async function updatePortalAccessKeyStatus(
  accountId: S3AccountSelector,
  accessKeyId: string,
  active: boolean,
): Promise<PortalAccessKey> {
  const { data } = await client.put<PortalAccessKey>(
    `/portal/access-keys/${encodeURIComponent(accessKeyId)}/status`,
    { active },
    { params: withS3AccountParam(undefined, accountId) },
  );
  return data;
}

export async function deletePortalAccessKey(
  accountId: S3AccountSelector,
  accessKeyId: string,
): Promise<void> {
  await client.delete(`/portal/access-keys/${encodeURIComponent(accessKeyId)}`, {
    params: withS3AccountParam(undefined, accountId),
  });
}
